File: Leetcode 3901-4000/3934/3934.java
// Leetcode 3934: Smallest Unique Subarray
// https://leetcode.com/problems/smallest-unique-subarray/
// Solved on 2nd of June, 2026
import java.util.Arrays;

class Solution {
    /**
     * Finds the length of the smallest subarray that appears exactly once in the given array.
     * 
     * @param nums An array of integers.
     * @return The length of the smallest unique subarray.
     */
    public int smallestUniqueSubarray(int[] nums) {
        int n = nums.length;
        int low = 1;
        int high = n;
        int ans = n;
        
        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (hasUnique(mid, nums)) {
                ans = mid;
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        
        return ans;
    }

    private boolean hasUnique(int mid, int[] nums) {
        int n = nums.length;
        long[] hashes = new long[n - mid + 1];
        long h1 = 0;
        long h2 = 0;
        long p1 = 1;
        long p2 = 1;
        long m1 = 1000000007;
        long m2 = 1000000009;
        long b1 = 13331;
        long b2 = 100003;
        
        for (int i = 0; i < mid; i++) {
            h1 = (h1 * b1 + nums[i]) % m1;
            h2 = (h2 * b2 + nums[i]) % m2;
            if (i < mid - 1) {
                p1 = (p1 * b1) % m1;
                p2 = (p2 * b2) % m2;
            }
        }
        
        hashes[0] = (h1 << 32) | h2;
        
        for (int i = mid; i < n; i++) {
            h1 = (h1 - (nums[i - mid] * p1) % m1 + m1) % m1;
            h1 = (h1 * b1 + nums[i]) % m1;
            h2 = (h2 - (nums[i - mid] * p2) % m2 + m2) % m2;
            h2 = (h2 * b2 + nums[i]) % m2;
            hashes[i - mid + 1] = (h1 << 32) | h2;
        }
        
        Arrays.sort(hashes);
        
        int count = 1;
        for (int i = 1; i < hashes.length; i++) {
            if (hashes[i] == hashes[i - 1]) {
                count++;
            } else {
                if (count == 1) {
                    return true;
                }
                count = 1;
            }
        }
        
        return count == 1;
    }
}

File: Leetcode 3901-4000/3934/3934.py
# Leetcode 3934: Smallest Unique Subarray
# https://leetcode.com/problems/smallest-unique-subarray/
# Solved on 2nd of June, 2026
class Solution:
    def smallestUniqueSubarray(self, nums: list[int]) -> int:
        """
        Finds the length of the smallest subarray that appears exactly once in the given list.

        Args:
            nums (list[int]): The input list of integers.
        Returns:
            int: The length of the shortest unique subarray.
        """
        leftBound = 1
        rightBound = len(nums)
        minUniqueLength = len(nums)
        baseVal = 100003
        modVal = (1 << 61) - 1

        while leftBound <= rightBound:
            midVal = (leftBound + rightBound) // 2
            currentHash = 0
            powerVal = 1

            for i in range(midVal):
                currentHash = (currentHash * baseVal + nums[i]) % modVal
                if i > 0:
                    powerVal = (powerVal * baseVal) % modVal

            hashCounts = {}
            hashCounts[currentHash] = 1

            for i in range(midVal, len(nums)):
                currentHash = (currentHash - nums[i - midVal] * powerVal) % modVal
                currentHash = (currentHash * baseVal + nums[i]) % modVal

                if currentHash in hashCounts:
                    hashCounts[currentHash] += 1
                else:
                    hashCounts[currentHash] = 1

            foundUnique = False
            for countVal in hashCounts.values():
                if countVal == 1:
                    foundUnique = True
                    break

            if foundUnique:
                minUniqueLength = midVal
                rightBound = midVal - 1
            else:
                leftBound = midVal + 1

        return minUniqueLength